custom/ilsgateway/tanzania/reports/unrecognized_sms.py

```diff
@@ -0,0 +1,165 @@
+from django.utils.translation import ugettext_noop
+from casexml.apps.case.models import CommCareCase
+from corehq.apps.hqwebapp.doc_info import get_doc_info
+from corehq.apps.locations.models import SQLLocation
+from corehq.apps.reports.datatables import DataTablesHeader, DataTablesColumn
+from corehq.apps.reports.filters.dates import DatespanFilter
+from corehq.apps.reports.filters.fixtures import AsyncLocationFilter
+from corehq.apps.reports.generic import GenericTabularReport
+from corehq.apps.reports.standard import CustomProjectReport, ProjectReportParametersMixin, DatespanMixin
+from corehq.apps.reports.util import format_datatables_data
+from corehq.apps.sms.models import WORKFLOW_DEFAULT, SMSLog, INCOMING, OUTGOING
+from corehq.apps.users.models import CouchUser
+from corehq.const import SERVER_DATETIME_FORMAT
+from corehq.util.timezones.conversions import ServerTime
+from dimagi.utils.parsing import json_format_datetime
+from django.utils.translation import ugettext as _
+import settings
+
+
+# copy paste from: https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/reports/standard/sms.py
+# we can't extend report by MessageLogReport.
+def _fmt(val):
+    if val is None:
+        return format_datatables_data("-", "-")
+    else:
+        return format_datatables_data(val, val)
+
+
+class UnrecognizedSMSReport(CustomProjectReport, ProjectReportParametersMixin,
+                            GenericTabularReport, DatespanMixin):
+    name = ugettext_noop('Unrecognized SMS')
+    slug = 'unrecognized_sms'
+    fields = [AsyncLocationFilter, DatespanFilter]
+    exportable = True
+
+    def _fmt_timestamp(self, timestamp):
+        return self.table_cell(
+            timestamp,
+            timestamp.strftime(SERVER_DATETIME_FORMAT),
+        )
+
+    def _fmt_contact_link(self, msg, doc_info):
+        if doc_info:
+            username, contact_type, url = (doc_info.display,
+                doc_info.type_display, doc_info.link)
+        else:
+            username, contact_type, url = (None, None, None)
+        username = username or "-"
+        contact_type = contact_type or _("Unknown")
+        if url:
+            ret = self.table_cell(username, '<a href="%s">%s</a>' % (url, username))
+        else:
+            ret = self.table_cell(username, username)
+        ret['raw'] = "|||".join([username, contact_type,
+            msg.couch_recipient or ""])
+        return ret
+
+    def get_recipient_info(self, message, contact_cache):
```

Comment:
this looks like a good fit for `quickcache` since the information is expected to be valid ~forever: https://github.com/dimagi/commcare-hq/blob/master/corehq/util/quickcache.py#L196